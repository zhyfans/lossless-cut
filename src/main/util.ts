import os from 'node:os';

export const platform = os.platform();
export const arch = os.arch();

export const isWindows = platform === 'win32';
export const isMac = platform === 'darwin';
export const isLinux = platform === 'linux';
