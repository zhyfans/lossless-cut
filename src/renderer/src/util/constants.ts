// anything more than this will probably cause the UI to become unusably slow
export const maxSegmentsAllowed = 2000;

export const ffmpegExtractWindow = 60;

export const zoomMax = 2 ** 14;

export const rightBarWidth = 200;
export const leftBarWidth = 240;
