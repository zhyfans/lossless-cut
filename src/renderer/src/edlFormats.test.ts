import { it, describe, expect, test } from 'vitest';

import { parseSrtToSegments, formatSrt, parseYouTube, formatYouTube, parseMplayerEdl, parseXmeml, parseFcpXml, parseCsv, parseCsvTime, getFrameValParser, formatCsvFrames, getFrameCountRaw, parsePbf, parseDvAnalyzerSummaryTxt, parseCutlist, parseGpsLine } from './edlFormats';
import { readFixture, readFixtureBinary } from './test/util';


const expectYouTube1 = [
  { start: 0, end: 1, name: '00:01 Test 1' },
  { start: 1, end: 2, name: '"Test 2":' },
  { start: 2, end: 4, name: '00:57 double' },
  { start: 4, end: 5, name: '' },
  { start: 5, end: 6, name: '' },
  { start: 6, end: 6.01, name: '6 label' },
  { start: 6.01, end: 61, name: '6.01 label' },
  { start: 61, end: 61.012, name: 'Test 3' },
  { start: 61.012, end: 62.012, name: 'Test 6' },
  { start: 62.012, end: 132, name: 'Test 7' },
  { start: 132, end: 3661.012, name: 'Integer' },
  { start: 3661.012, end: 10074, name: 'Test - 4' },
  { start: 10074, end: undefined, name: 'Short - hour and hyphen' },
];

describe('parseYouTube', () => {
  it('parses different cases', () => {
    const str = `
  Jump to chapters:
  0:00 00:01 Test 1
  00:01 "Test 2":
  00:02 00:57 double
  00:01:01 Test 3
  01:01:01.012 Test - 4
  00:01:01.012 Test 5
     01:01.012 Test 6
    :01:02.012 Test 7
  2:47:54 - Short - hour and hyphen
  00:57:01.0123 Invalid 2
  00:57:01. Invalid 3
  01:15:: Invalid 4
  0132 Integer
  00:03
  00:04     
  00:05 
  6 6 label
  6.01 6.01 label
  `;
    const edl = parseYouTube(str);
    expect(edl).toEqual(expectYouTube1);
  });

  it('eol', () => {
    const str = ' 00:00 Test 1\n00:01 Test 2';
    const edl = parseYouTube(str);
    expect(edl).toEqual([
      { start: 0, end: 1, name: 'Test 1' },
      { start: 1, end: undefined, name: 'Test 2' },
    ]);
  });

  it('colon after time', () => {
    const str = ' 00:00: Test 1';
    const edl = parseYouTube(str);
    expect(edl).toEqual([
      { start: 0, end: undefined, name: 'Test 1' },
    ]);
  });
});

it('formatYouTube', () => {
  expect(formatYouTube([
    { start: 1, end: 2, name: 'Label ðŸŽ‰' },
    { start: 3, end: 5 },
    { start: 10000, end: 10001, name: '' },
  ]).split('\n')).toEqual([
    '0:01 Label ðŸŽ‰',
    '0:03',
    '2:46:40',
  ]);
  expect(formatYouTube([
    { start: 0, end: 100 },
  ]).split('\n')).toEqual([
    '0:00',
  ]);
});

it('formatYouTube 2', () => {
  expect(formatYouTube(expectYouTube1).split('\n')).toEqual([
    '0:00 00:01 Test 1',
    '0:01 "Test 2":',
    '0:02 00:57 double',
    '0:04',
    '0:05',
    '0:06 6 label',
    '0:06 6.01 label',
    '1:01 Test 3',
    '1:01 Test 6',
    '1:02 Test 7',
    '2:12 Integer',
    '1:01:01 Test - 4',
    '2:47:54 Short - hour and hyphen',
  ]);
});

// https://kodi.wiki/view/Edit_decision_list
// http://www.mplayerhq.hu/DOCS/HTML/en/edl.html
it('parseMplayerEdl', async () => {
  // TODO support more durations (frames and timestamps):
  /*
  const str = `\
5.3     7.1     0
15      16.7    1
7:00    13:42   3
1       4:15.3  2
12:00.1         2
`;
  const str = `\
#127   #170    0
#360   #400    1
#10080 #19728  3
#1     #6127   2
#17282         2
`;
*/

  const str = await readFixture('mplayer.edl');

  expect(await parseMplayerEdl(str)).toEqual([
    { start: 0,
      end: 5.3,
      name: 'Cut',
      tags: { mplayerEdlType: '0' },
    },
    { start: 7.1,
      end: undefined,
      name: 'Cut',
      tags: { mplayerEdlType: '0' },
    },
    {
      end: 16.7,
      start: 15,
      name: 'Mute',
      tags: { mplayerEdlType: '1' },
    },
    {
      end: 255.3,
      start: 1,
      name: 'Scene Marker',
      tags: { mplayerEdlType: '2' },
    },
    {
      end: 822,
      start: 420,
      name: 'Commercial Break',
      tags: { mplayerEdlType: '3' },
    },
  ]);
});

it('parseMplayerEdl, starting at 0', async () => {
  const str2 = '  0   1.1    0\n';

  expect(await parseMplayerEdl(str2)).toEqual([{
    start: 1.1,
    end: undefined,
    name: 'Cut',
    tags: {
      mplayerEdlType: '0',
    },
  }]);
});

it('parses xmeml 1', async () => {
  expect(parseXmeml(await readFixture('Final Cut Pro XMEML.xml'))).toMatchSnapshot();
});

it('parses xmeml 2', async () => {
  expect(parseXmeml(await readFixture('Final Cut Pro XMEML 2.xml'))).toMatchSnapshot();
});

// see https://github.com/mifi/lossless-cut/issues/1195
it('parses xmeml - with multiple tracks', async () => {
  expect(parseXmeml(await readFixture('Final Cut Pro XMEML 3.xml'))).toMatchSnapshot();
});

// see https://github.com/mifi/lossless-cut/issues/1195
it('parses fcpxml 1.9', async () => {
  expect(await parseFcpXml(await readFixture('FCPXML_1_9.fcpxml'))).toMatchSnapshot();
});

// https://github.com/mifi/lossless-cut/issues/1024
const csvFramesStr = `\
0,155,EP106_SQ010_SH0010
156,251,EP106_SQ010_SH0020
252,687,EP106_SQ010_SH0030
688,747,EP106_SQ020_SH0010
`;

it('parses csv with frames', async () => {
  const fps = 30;
  const parsed = await parseCsv(csvFramesStr, getFrameValParser(fps));

  expect(parsed).toEqual([
    { end: 5.166666666666667, name: 'EP106_SQ010_SH0010', start: 0 },
    { end: 8.366666666666667, name: 'EP106_SQ010_SH0020', start: 5.2 },
    { end: 22.9, name: 'EP106_SQ010_SH0030', start: 8.4 },
    { end: 24.9, name: 'EP106_SQ020_SH0010', start: 22.933333333333334 },
  ]);

  const formatted = await formatCsvFrames({
    cutSegments: parsed,
    getFrameCount: (sec) => getFrameCountRaw(fps, sec),
  });
  expect(formatted).toEqual(csvFramesStr);
});

const csvTimestampStr = `\
00:01:54.612,00:03:09.053,A
  00:05:00.448,00:07:56.194,B  
00:09:27.075,00:11:44.264,C
0,1,D
1.01,1.99,E
0:2,0:3,F
`;

it('parses csv with timestamps', async () => {
  const parsed = await parseCsv(csvTimestampStr, parseCsvTime);

  expect(parsed).toEqual([
    { end: 189.053, name: 'A', start: 114.612 },
    { end: 476.194, name: 'B', start: 300.448 },
    { end: 704.264, name: 'C', start: 567.075 },
    { end: 1, name: 'D', start: 0 },
    { end: 1.99, name: 'E', start: 1.01 },
    { start: 2, name: 'F', end: 3 },
  ]);
});

const cutlistStr = `
[General]
Application=SomeApplication.exe
Version=0.0.0.1
FramesPerSecond=25
IntendedCutApplicationName=SomeApplication
IntendedCutApplication=SomeApplication.exe
VDUseSmartRendering=1
IntendedCutApplicationVersion=1.7.8
comment1=The following parts of the movie will be kept, the rest will be cut out.
comment2=All values are given in seconds.
NoOfCuts=2
ApplyToFile=Some_File_Name.avi
OriginalFileSizeBytes=123456

[Cut0]
Start=849.12
StartFrame=21228
Duration=1881.84
DurationFrames=47046

[Cut1]
Start=3147.72
StartFrame=78693
Duration=944.6
DurationFrames=23615

[Info]
Author=AuthorName
RatingByAuthor=0
EPGError=0
ActualContent=
MissingBeginning=0
MissingEnding=0
MissingVideo=0
MissingAudio=0
OtherError=0
OtherErrorDescription=
SuggestedMovieName=
UserComment=cutted with XXXX

[Meta]
CutlistId=12345
GeneratedOn=1900-01-01 00:00:01
GeneratedBy=cutlist v0.0.0
`;

it('parses cutlist', async () => {
  const parsed = await parseCutlist(cutlistStr);

  expect(parsed).toEqual([
    { end: 2730.96, name: 'Cut 0', start: 849.12 },
    { end: 4092.32, name: 'Cut 1', start: 3147.72 },
  ]);
});

it('parses pbf', async () => {
  expect(parsePbf(await readFixtureBinary('test1.pbf'))).toMatchSnapshot();
  expect(parsePbf(await readFixtureBinary('test2.pbf'))).toMatchSnapshot();
  expect(parsePbf(await readFixtureBinary('test3.pbf'))).toMatchSnapshot();
  expect(parsePbf(await readFixtureBinary('potplayer bookmark format utf16le issue 867.pbf'))).toMatchSnapshot();
});

it('parses srt', async () => {
  expect(parseSrtToSegments(await readFixture('sample.srt'))).toMatchSnapshot();
});

it('format srt', async () => {
  expect(formatSrt(parseSrtToSegments(await readFixture('sample.srt')))).toMatchSnapshot();
});

// https://github.com/mifi/lossless-cut/issues/1664
it('parses DV Analyzer Summary.txt', async () => {
  expect(parseDvAnalyzerSummaryTxt(await readFixture('DV Analyzer Summary.txt', 'utf8'))).toMatchSnapshot();
});

test('parseGpsLine', () => {
  expect(parseGpsLine('F/2.8, SS 776.89, ISO 100, EV -1.0, GPS (15.0732, 67.9771, 19), D 67.78m, H 20.30m, H.S 1.03m/s, V.S 0.00m/s')).toMatchSnapshot();
  // https://github.com/mifi/lossless-cut/issues/2072#issuecomment-2325755148
  expect(parseGpsLine('F/2.8, SS 678.52, ISO 100, EV 0, DZOOM 1.000, GPS (-3.9130, 56.5019, 26), D 0.57m, H 102.40m, H.S 0.00m/s, V.S -0.00m/s')).toMatchSnapshot();
});
