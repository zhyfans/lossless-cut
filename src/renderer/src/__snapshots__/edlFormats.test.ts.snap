// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`format srt 1`] = `
"1
00:00:02,000 --> 00:00:06,000
First subtitle

2
00:00:28,967 --> 01:30:30,950
Subtitle 2 line 1
Subtitle 2 line 2
"
`;

exports[`parseGpsLine 1`] = `
{
  "alt": 19,
  "distance": 67.78,
  "dzoom": undefined,
  "ev": -1,
  "f": "F/2.8",
  "height": 20.3,
  "horizontalSpeed": 1.03,
  "iso": 100,
  "lat": 15.0732,
  "lng": 67.9771,
  "ss": 776.89,
  "verticalSpeed": 0,
}
`;

exports[`parseGpsLine 2`] = `
{
  "alt": 26,
  "distance": 0.57,
  "dzoom": 1,
  "ev": 0,
  "f": "F/2.8",
  "height": 102.4,
  "horizontalSpeed": 0,
  "iso": 100,
  "lat": -3.913,
  "lng": 56.5019,
  "ss": 678.52,
  "verticalSpeed": -0,
}
`;

exports[`parses DV Analyzer Summary.txt 1`] = `
[
  {
    "end": 60.4,
    "name": "XXXX-XX-XX 00:00:00.000",
    "start": 0,
    "tags": {
      "recordedEnd": "XXXX-XX-XX XX:XX:XX:XX",
      "recordedStart": "XXXX-XX-XX 00:00:00.000",
    },
  },
  {
    "end": 485.08,
    "name": "XXXX-XX-XX XX:XX:XX:XX",
    "start": 60.4,
    "tags": {
      "recordedEnd": "2001-12-31 23:22:09",
      "recordedStart": "XXXX-XX-XX XX:XX:XX:XX",
    },
  },
  {
    "end": 1010.68,
    "name": "2001-12-31 23:28:13",
    "start": 485.08,
    "tags": {
      "recordedEnd": "2002-01-01 19:34:38",
      "recordedStart": "2001-12-31 23:28:13",
    },
  },
  {
    "end": 1235.32,
    "name": "2002-01-01 13:31:24",
    "start": 1010.68,
    "tags": {
      "recordedEnd": "2002-01-01 22:03:01",
      "recordedStart": "2002-01-01 13:31:24",
    },
  },
  {
    "end": 1575.04,
    "name": "2002-01-02 14:27:10",
    "start": 1235.32,
    "tags": {
      "recordedEnd": "2002-01-02 15:48:55",
      "recordedStart": "2002-01-02 14:27:10",
    },
  },
  {
    "end": 1575.08,
    "name": "2002-01-02 22:30:22",
    "start": 1575.04,
    "tags": {
      "recordedEnd": "2002-01-02 22:30:22",
      "recordedStart": "2002-01-02 22:30:22",
    },
  },
  {
    "end": 1575.16,
    "name": "2002-01-02 22:30:22",
    "start": 1575.08,
    "tags": {
      "recordedEnd": "2002-01-02 22:30:22",
      "recordedStart": "2002-01-02 22:30:22",
    },
  },
  {
    "end": 1575.24,
    "name": "2002-01-02 22:30:22",
    "start": 1575.16,
    "tags": {
      "recordedEnd": "2002-01-05 10:57:51",
      "recordedStart": "2002-01-02 22:30:22",
    },
  },
  {
    "end": 1919.44,
    "name": "2002-01-05 10:57:51",
    "start": 1575.24,
    "tags": {
      "recordedEnd": "2002-01-05 11:36:20",
      "recordedStart": "2002-01-05 10:57:51",
    },
  },
  {
    "end": 2075.88,
    "name": "2002-01-05 13:18:43",
    "start": 1919.44,
    "tags": {
      "recordedEnd": "2002-01-05 14:04:19",
      "recordedStart": "2002-01-05 13:18:43",
    },
  },
  {
    "end": 2138.2,
    "name": "2002-01-05 16:39:22",
    "start": 2075.88,
    "tags": {
      "recordedEnd": "2002-01-05 22:51:40",
      "recordedStart": "2002-01-05 16:39:22",
    },
  },
  {
    "end": 2138.76,
    "name": "2002-01-05 16:40:24",
    "start": 2138.2,
    "tags": {
      "recordedEnd": "2002-01-05 16:40:25",
      "recordedStart": "2002-01-05 16:40:24",
    },
  },
  {
    "end": 2217.32,
    "name": "2002-01-05 22:53:17",
    "start": 2138.76,
    "tags": {
      "recordedEnd": "2002-01-05 22:55:08",
      "recordedStart": "2002-01-05 22:53:17",
    },
  },
  {
    "end": 2269.96,
    "name": "2002-01-16 21:17:04",
    "start": 2217.32,
    "tags": {
      "recordedEnd": "2002-01-16 21:18:01",
      "recordedStart": "2002-01-16 21:17:04",
    },
  },
  {
    "end": 2332.64,
    "name": "2002-01-20 20:06:37",
    "start": 2269.96,
    "tags": {
      "recordedEnd": "2002-01-20 20:07:48",
      "recordedStart": "2002-01-20 20:06:37",
    },
  },
  {
    "end": 3009.12,
    "name": "2002-01-30 18:34:52",
    "start": 2332.64,
    "tags": {
      "recordedEnd": "2002-03-12 00:46:51",
      "recordedStart": "2002-01-30 18:34:52",
    },
  },
  {
    "end": 3596.48,
    "name": "2002-03-14 20:27:57",
    "start": 3009.12,
    "tags": {
      "recordedEnd": "2002-04-12 21:06:54",
      "recordedStart": "2002-03-14 20:27:57",
    },
  },
  {
    "end": 3622.08,
    "name": "2002-04-12 21:06:56",
    "start": 3596.48,
    "tags": {
      "recordedEnd": "2002-04-12 21:07:22",
      "recordedStart": "2002-04-12 21:06:56",
    },
  },
  {
    "end": 4305,
    "name": "2002-04-12 21:11:47",
    "start": 3622.08,
    "tags": {
      "recordedEnd": "2002-04-27 00:05:36",
      "recordedStart": "2002-04-12 21:11:47",
    },
  },
  {
    "end": 4307.12,
    "name": "2002-04-25 22:59:57",
    "start": 4305,
    "tags": {
      "recordedEnd": "2002-04-25 22:59:59",
      "recordedStart": "2002-04-25 22:59:57",
    },
  },
  {
    "end": 4357.68,
    "name": "2002-04-25 23:00:00",
    "start": 4307.12,
    "tags": {
      "recordedEnd": "2002-05-02 13:33:33",
      "recordedStart": "2002-04-25 23:00:00",
    },
  },
  {
    "end": 4359.72,
    "name": "2002-04-25 23:00:50",
    "start": 4357.68,
    "tags": {
      "recordedEnd": "2002-04-25 23:00:52",
      "recordedStart": "2002-04-25 23:00:50",
    },
  },
  {
    "end": 4660.44,
    "name": "2002-05-02 13:40:27",
    "start": 4359.72,
    "tags": {
      "recordedEnd": "2002-05-02 13:53:34",
      "recordedStart": "2002-05-02 13:40:27",
    },
  },
  {
    "end": undefined,
    "name": "2002-05-02 13:54:14",
    "start": 4660.44,
    "tags": {
      "recordedEnd": "2002-05-02 13:59:29",
      "recordedStart": "2002-05-02 13:54:14",
    },
  },
]
`;

exports[`parses fcpxml 1.9 1`] = `
[
  {
    "end": 13.84,
    "start": 0,
  },
  {
    "end": 25.04,
    "start": 13.84,
  },
  {
    "end": 25.2,
    "start": 25.04,
  },
  {
    "end": 55.44,
    "start": 25.2,
  },
  {
    "end": 68.96,
    "start": 55.44,
  },
  {
    "end": 81.83999999999999,
    "start": 68.96,
  },
  {
    "end": 95.12,
    "start": 81.84,
  },
  {
    "end": 95.16000000000001,
    "start": 95.12,
  },
  {
    "end": 95.28,
    "start": 95.16,
  },
  {
    "end": 112.72,
    "start": 95.28,
  },
  {
    "end": 140.4,
    "start": 112.72,
  },
  {
    "end": 147.36,
    "start": 140.4,
  },
  {
    "end": 170.88000000000002,
    "start": 147.36,
  },
  {
    "end": 176.72,
    "start": 170.88,
  },
  {
    "end": 196.07999999999998,
    "start": 176.72,
  },
  {
    "end": 232.96,
    "start": 196.08,
  },
  {
    "end": 265.92,
    "start": 232.96,
  },
  {
    "end": 281.76,
    "start": 265.92,
  },
  {
    "end": 308.96,
    "start": 281.76,
  },
  {
    "end": 321.68,
    "start": 308.96,
  },
  {
    "end": 353.2,
    "start": 321.68,
  },
  {
    "end": 376.08,
    "start": 353.2,
  },
  {
    "end": 405.68,
    "start": 376.08,
  },
  {
    "end": 412.16,
    "start": 405.68,
  },
  {
    "end": 423.28000000000003,
    "start": 412.16,
  },
  {
    "end": 430.71999999999997,
    "start": 423.28,
  },
  {
    "end": 439.76000000000005,
    "start": 430.72,
  },
  {
    "end": 456.64,
    "start": 439.76,
  },
  {
    "end": 492.71999999999997,
    "start": 456.64,
  },
  {
    "end": 511.44000000000005,
    "start": 492.72,
  },
  {
    "end": 522.24,
    "start": 511.44,
  },
  {
    "end": 524.32,
    "start": 522.24,
  },
  {
    "end": 543.2800000000001,
    "start": 524.32,
  },
  {
    "end": 571.04,
    "start": 543.28,
  },
  {
    "end": 576.3199999999999,
    "start": 571.04,
  },
  {
    "end": 612.48,
    "start": 576.32,
  },
  {
    "end": 626.16,
    "start": 612.48,
  },
  {
    "end": 626.24,
    "start": 626.16,
  },
  {
    "end": 665.04,
    "start": 626.24,
  },
  {
    "end": 694.0799999999999,
    "start": 665.04,
  },
  {
    "end": 762.96,
    "start": 694.08,
  },
  {
    "end": 784.8000000000001,
    "start": 762.96,
  },
  {
    "end": 804.56,
    "start": 784.8,
  },
  {
    "end": 813.52,
    "start": 804.56,
  },
  {
    "end": 831.16,
    "start": 813.52,
  },
  {
    "end": 928.72,
    "start": 831.16,
  },
  {
    "end": 941.0400000000001,
    "start": 928.72,
  },
  {
    "end": 946.3199999999999,
    "start": 941.04,
  },
  {
    "end": 968.8000000000001,
    "start": 946.32,
  },
  {
    "end": 982.16,
    "start": 968.8,
  },
  {
    "end": 1111.2,
    "start": 982.16,
  },
  {
    "end": 1117.04,
    "start": 1111.2,
  },
  {
    "end": 1142.6,
    "start": 1117.04,
  },
  {
    "end": 1142.6399999999999,
    "start": 1142.6,
  },
  {
    "end": 1142.68,
    "start": 1142.64,
  },
  {
    "end": 1180.88,
    "start": 1142.68,
  },
  {
    "end": 1193.2,
    "start": 1180.88,
  },
  {
    "end": 1212.72,
    "start": 1193.2,
  },
  {
    "end": 1233.68,
    "start": 1212.72,
  },
  {
    "end": 1263.76,
    "start": 1233.68,
  },
  {
    "end": 1288.56,
    "start": 1263.76,
  },
  {
    "end": 1294.12,
    "start": 1288.56,
  },
  {
    "end": 1294.1599999999999,
    "start": 1294.12,
  },
  {
    "end": 1294.2,
    "start": 1294.16,
  },
  {
    "end": 1294.24,
    "start": 1294.2,
  },
  {
    "end": 1294.28,
    "start": 1294.24,
  },
  {
    "end": 1294.36,
    "start": 1294.28,
  },
  {
    "end": 1294.4399999999998,
    "start": 1294.36,
  },
  {
    "end": 1294.48,
    "start": 1294.44,
  },
  {
    "end": 1306.84,
    "start": 1294.48,
  },
  {
    "end": 1306.8799999999999,
    "start": 1306.84,
  },
  {
    "end": 1306.92,
    "start": 1306.88,
  },
  {
    "end": 1306.96,
    "start": 1306.92,
  },
  {
    "end": 1339.04,
    "start": 1306.96,
  },
  {
    "end": 1360.24,
    "start": 1339.04,
  },
  {
    "end": 1365.92,
    "start": 1360.24,
  },
  {
    "end": 1382.88,
    "start": 1365.92,
  },
  {
    "end": 1402.0800000000002,
    "start": 1382.88,
  },
  {
    "end": 1417.52,
    "start": 1402.08,
  },
  {
    "end": 1428.16,
    "start": 1417.52,
  },
  {
    "end": 1480.0800000000002,
    "start": 1428.16,
  },
  {
    "end": 1496.24,
    "start": 1480.08,
  },
  {
    "end": 1526.64,
    "start": 1496.24,
  },
  {
    "end": 1532.24,
    "start": 1526.64,
  },
  {
    "end": 1542.24,
    "start": 1532.24,
  },
  {
    "end": 1554.32,
    "start": 1542.24,
  },
  {
    "end": 1565.12,
    "start": 1554.32,
  },
  {
    "end": 1595.4399999999998,
    "start": 1565.12,
  },
  {
    "end": 1602.3200000000002,
    "start": 1595.44,
  },
  {
    "end": 1612.32,
    "start": 1602.32,
  },
  {
    "end": 1648.24,
    "start": 1612.32,
  },
  {
    "end": 1667.76,
    "start": 1648.24,
  },
  {
    "end": 1691.92,
    "start": 1667.76,
  },
  {
    "end": 1705.76,
    "start": 1691.92,
  },
  {
    "end": 1711.12,
    "start": 1705.76,
  },
  {
    "end": 1727,
    "start": 1711.12,
  },
  {
    "end": 1750,
    "start": 1727,
  },
  {
    "end": 1778,
    "start": 1750,
  },
  {
    "end": 1787.6,
    "start": 1778,
  },
  {
    "end": 1794.9599999999998,
    "start": 1787.6,
  },
  {
    "end": 1836.32,
    "start": 1794.96,
  },
  {
    "end": 1905.84,
    "start": 1836.32,
  },
  {
    "end": 1905.8799999999999,
    "start": 1905.84,
  },
  {
    "end": 1909.1200000000001,
    "start": 1905.88,
  },
  {
    "end": 1936.56,
    "start": 1909.12,
  },
  {
    "end": 1941.84,
    "start": 1936.56,
  },
  {
    "end": 1954.48,
    "start": 1941.84,
  },
  {
    "end": 1968.24,
    "start": 1954.48,
  },
  {
    "end": 1972,
    "start": 1968.24,
  },
  {
    "end": 1982.56,
    "start": 1972,
  },
  {
    "end": 1994.56,
    "start": 1982.56,
  },
  {
    "end": 1997.12,
    "start": 1994.56,
  },
  {
    "end": 2164.3599999999997,
    "start": 1997.12,
  },
  {
    "end": 2180.4,
    "start": 2164.36,
  },
  {
    "end": 2199.28,
    "start": 2180.4,
  },
  {
    "end": 2212,
    "start": 2199.28,
  },
  {
    "end": 2237.76,
    "start": 2212,
  },
  {
    "end": 2316.88,
    "start": 2237.76,
  },
  {
    "end": 2332.2400000000002,
    "start": 2316.88,
  },
  {
    "end": 2412.64,
    "start": 2332.24,
  },
  {
    "end": 2672.56,
    "start": 2412.64,
  },
  {
    "end": 2723.24,
    "start": 2672.56,
  },
  {
    "end": 2723.2799999999997,
    "start": 2723.24,
  },
  {
    "end": 2723.36,
    "start": 2723.28,
  },
  {
    "end": 2723.4,
    "start": 2723.36,
  },
  {
    "end": 2725.04,
    "start": 2723.4,
  },
  {
    "end": 2725.08,
    "start": 2725.04,
  },
  {
    "end": 2725.12,
    "start": 2725.08,
  },
  {
    "end": 2725.44,
    "start": 2725.12,
  },
  {
    "end": 2787.2000000000003,
    "start": 2725.44,
  },
  {
    "end": 2806.56,
    "start": 2787.2,
  },
  {
    "end": 2814.56,
    "start": 2806.56,
  },
  {
    "end": 2852.4,
    "start": 2814.56,
  },
  {
    "end": 2864.8,
    "start": 2852.4,
  },
  {
    "end": 2875.28,
    "start": 2864.8,
  },
  {
    "end": 2886.6400000000003,
    "start": 2875.28,
  },
  {
    "end": 2898.3199999999997,
    "start": 2886.64,
  },
  {
    "end": 2942.2400000000002,
    "start": 2898.32,
  },
  {
    "end": 2996,
    "start": 2942.24,
  },
  {
    "end": 3000.88,
    "start": 2996,
  },
  {
    "end": 3026.6400000000003,
    "start": 3000.88,
  },
  {
    "end": 3055.92,
    "start": 3026.64,
  },
  {
    "end": 3062.7200000000003,
    "start": 3055.92,
  },
  {
    "end": 3088.8799999999997,
    "start": 3062.72,
  },
  {
    "end": 3095.76,
    "start": 3088.88,
  },
  {
    "end": 3111.36,
    "start": 3095.76,
  },
  {
    "end": 3118.1600000000003,
    "start": 3111.36,
  },
  {
    "end": 3125.8399999999997,
    "start": 3118.16,
  },
  {
    "end": 3131.28,
    "start": 3125.84,
  },
  {
    "end": 3139.28,
    "start": 3131.28,
  },
  {
    "end": 3174.1600000000003,
    "start": 3139.28,
  },
  {
    "end": 3177.68,
    "start": 3174.16,
  },
  {
    "end": 3177.72,
    "start": 3177.68,
  },
  {
    "end": 3193.12,
    "start": 3177.72,
  },
  {
    "end": 3218.24,
    "start": 3193.12,
  },
  {
    "end": 3235.7599999999998,
    "start": 3218.24,
  },
  {
    "end": 3250,
    "start": 3235.76,
  },
  {
    "end": 3263.76,
    "start": 3250,
  },
  {
    "end": 3267.28,
    "start": 3263.76,
  },
  {
    "end": 3272.96,
    "start": 3267.28,
  },
  {
    "end": 3335.92,
    "start": 3272.96,
  },
  {
    "end": 3499.76,
    "start": 3335.92,
  },
  {
    "end": 3640.7200000000003,
    "start": 3499.76,
  },
  {
    "end": 3655.9599999999996,
    "start": 3640.72,
  },
  {
    "end": 3656.04,
    "start": 3655.96,
  },
  {
    "end": 3668.16,
    "start": 3656.04,
  },
  {
    "end": 3744.7999999999997,
    "start": 3668.16,
  },
  {
    "end": 3770.8,
    "start": 3744.8,
  },
  {
    "end": 3783.6000000000004,
    "start": 3770.8,
  },
  {
    "end": 3793.68,
    "start": 3783.6,
  },
  {
    "end": 3814.64,
    "start": 3793.68,
  },
  {
    "end": 3835.04,
    "start": 3814.64,
  },
  {
    "end": 3843.92,
    "start": 3835.04,
  },
]
`;

exports[`parses pbf 1`] = `
[
  {
    "end": 189.053,
    "name": "label 1",
    "start": 104.612,
  },
  {
    "end": 476.194,
    "name": "label 2",
    "start": 300.448,
  },
  {
    "end": 704.264,
    "name": "label 3",
    "start": 567.075,
  },
  {
    "end": 855.455,
    "name": "label 4",
    "start": 704.264,
  },
]
`;

exports[`parses pbf 2`] = `
[
  {
    "end": 189.053,
    "name": "label 1",
    "start": 104.612,
  },
  {
    "end": 300.448,
    "name": "label 2",
    "start": 189.053,
  },
  {
    "end": 476.194,
    "name": "label 3",
    "start": 300.448,
  },
  {
    "end": 567.075,
    "name": "label 4",
    "start": 476.194,
  },
  {
    "end": 704.264,
    "name": "label 5",
    "start": 567.075,
  },
  {
    "end": 855.455,
    "name": "label 6",
    "start": 704.264,
  },
]
`;

exports[`parses pbf 3`] = `
[
  {
    "end": 189.053,
    "name": "A",
    "start": 104.612,
  },
  {
    "end": 476.194,
    "name": "B",
    "start": 300.448,
  },
  {
    "end": 704.264,
    "name": "C",
    "start": 567.075,
  },
]
`;

exports[`parses pbf 4`] = `
[
  {
    "end": 189.053,
    "name": "A",
    "start": 104.612,
  },
  {
    "end": 476.194,
    "name": "B",
    "start": 300.448,
  },
]
`;

exports[`parses srt 1`] = `
[
  {
    "end": 6,
    "name": "First subtitle",
    "start": 2,
    "tags": {
      "index": "1",
    },
  },
  {
    "end": 5430.95,
    "name": "Subtitle 2 line 1
Subtitle 2 line 2",
    "start": 28.967,
    "tags": {
      "index": "2",
    },
  },
]
`;

exports[`parses xmeml - with multiple tracks 1`] = `
[
  {
    "end": 11,
    "start": 0,
  },
  {
    "end": 4.08,
    "start": 0.96,
  },
  {
    "end": 3.88,
    "start": 1.12,
  },
  {
    "end": 4.32,
    "start": 0.84,
  },
  {
    "end": 3.36,
    "start": 0.36,
  },
  {
    "end": 6.24,
    "start": 0.36,
  },
  {
    "end": 4.52,
    "start": 0.24,
  },
  {
    "end": 2.48,
    "start": 0.48,
  },
  {
    "end": 2.36,
    "start": 0.48,
  },
  {
    "end": 1.36,
    "start": 0.36,
  },
  {
    "end": 1.36,
    "start": 0.36,
  },
  {
    "end": 1.04,
    "start": 0.36,
  },
]
`;

exports[`parses xmeml 1 1`] = `
[
  {
    "end": 8.56,
    "start": 0,
  },
  {
    "end": 13.12,
    "start": 8.56,
  },
  {
    "end": 19.12,
    "start": 13.12,
  },
  {
    "end": 27.96,
    "start": 19.12,
  },
  {
    "end": 39.08,
    "start": 27.96,
  },
  {
    "end": 49.36,
    "start": 39.08,
  },
]
`;

exports[`parses xmeml 2 1`] = `
[
  {
    "end": 2033.5666666666666,
    "start": 5.1,
  },
  {
    "end": 3839.2833333333333,
    "start": 3017.7,
  },
  {
    "end": 4686.366666666667,
    "start": 3862.9333333333334,
  },
  {
    "end": 6310.016666666666,
    "start": 5324.066666666667,
  },
  {
    "end": 6959.516666666666,
    "start": 6928.15,
  },
  {
    "end": 8081.116666666667,
    "start": 6985.9,
  },
  {
    "end": 9750.05,
    "start": 8728.066666666668,
  },
  {
    "end": 10545.116666666667,
    "start": 9793.316666666668,
  },
  {
    "end": 10864.883333333333,
    "start": 10740.633333333333,
  },
]
`;
